pandas/core/strings.py

```diff
@@ -107,50 +112,117 @@ def cat_safe(list_of_columns: List, sep: str):
     return result
 
 
-def _na_map(f, arr, na_result=np.nan, dtype=object):
-    # should really _check_ for NA
-    return _map(f, arr, na_mask=True, na_value=na_result, dtype=dtype)
+def _masked_map(f, arr: ArrayLike, result_dtype, na_result=np.nan, method=None):
+    """
+    Map a method over valid values in an array.
 
+    The callable `f` will *not* be applied to missing values. Instead,
+    it is only applied to valid values. Missing values will be set to
+    `na_result` in the output.
 
-def _map(f, arr, na_mask=False, na_value=np.nan, dtype=object):
-    if not len(arr):
-        return np.ndarray(0, dtype=dtype)
+    Parameters
+    ----------
+    f : Callable
+        Should be a single-argument callable.
+    arr : ArrayLike
+        Array of values to apply the function to.
+    result_dtype : {"int", "bool", "string", "object"}
+        The output dtype. For most methods (e.g. str.upper, str.isint)
+        this can be known statically. For some, (e.g. str.get) we need
+        to infer the result dtype based on the values. See `method`.
+    na_result : Any
+        The result value to use for missing values.
+    method : str, optional.
+        Most methods should *not* pass this; it's unnecessary. It's
+        only necessary for `.str` methods that don't really operate
+        (just) on strings. For this reason, it's only relevant for
+        object-dtype input, and is completely unnecessary for
+        StringArray.
+    """
+    assert result_dtype in {"int", "bool", "string", "object"}
+    arr = extract_array(arr, extract_numpy=True)
+    mask = isna(arr)
+
+    if is_extension_array_dtype(arr.dtype):
+        # This is our nice, future world.  We *always* know the output
+        # dtype based just on `result_dtype`, no need to look at
+        # the values.
+        notna = ~mask
+        anyna = np.any(mask)
+
+        if result_dtype == "int":
+            result_dtype = pandas_dtype("Int64")
+            # TODO: Avoid this object allocation
```

Comment:
I think this optimization would be something like

```python
result = np.zeros(dtype="int64")
result[notna] = lib.map_infer(np.asarray(arr[notna]), f)
return IntegerArray(result, mask)
```

which skips the object-dtype ndarray allocation.